### PR TITLE
feat(*): add optional wasm filter config validation

### PR DESCRIPTION
**Pull request by flrgh**
Changes:
  * introduces a new `json` field type with JSON schema-powered validation
  * implement discovery and import of `{{ filter }}.meta.json` files for providing Kong with Wasm filter metadata (such as configuration schema)
  * Add a new `json_config` field to `filter_chains.filters[]` which is optionally validated against JSON schema from the filter metadata file (if present)

[KAG-662](https://konghq.atlassian.net/browse/KAG-662)

[KAG-662]: https://konghq.atlassian.net/browse/KAG-662?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: kong/db/schema/entities/filter_chains.lua
Comment (flrgh) [inline kong/db/schema/entities/filter_chains.lua:63]: This validation function is still in the "just make the tests pass" stage because I didn't want to tinker with it endlessly. The performance outlook of this is not good, especially given that this is per-filter validation, so it will execute in a loop (once for each of `filter_chains[].filters`).

Seeking some pointers on how to best optimize this (best referring to the maintainability moreso than the performance). Should I just use `kong.cache:get()` to memoize `select()` ops? Should I try to add a one-off JOIN query to the DAO so that we can reduce these two `select()`s into one?

None of the other code in `kong/db/*` seems to depend upon `kong.cache`, so it feels like it might be bad practice to reach for memoization here.

Alternatively, maybe it's fine to leave as is for the time being? This is a brand new feature _within_ another brand new feature (Wasm) which is still in tech preview, so I am open to the idea that it doesn't need to be fully-optimized yet.

File: kong/db/schema/entities/schemas.lua
Comment (flrgh) [inline kong/db/schema/entities/schemas.lua:38]: Note: because the full list of possible schema types is static and known at startup, we could easily rewrite this to use sub-schemas (each `type` is a sub-schema with its own `custom_validator`). Would that be preferable to using `custom_entity_check`?

File: kong/db/schema/entities/schemas.lua
Comment (flrgh) [inline kong/db/schema/entities/schemas.lua:16]: The name `schemas` is generic enough that it _might_ yield confusion and/or naming collisions. Should we consider calling it `user_schemas`?

Comment (hishamhm) [reply kong/db/schema/entities/schemas.lua:16]: (to clarify, I think the table should remain `schemas` and the schema names themselves should contain namespacing)

File: kong/db/schema/entities/filter_chains.lua
Comment (flrgh) [reply kong/db/schema/entities/filter_chains.lua:63]: I think something like that would be nice at wiping out all the lookup logic from this function. Something along the lines of...

```lua
local filter = {
  type = "record",
  fields = {
    { name    = { type = "string", required = true, one_of = wasm.filter_names,
                  err = "no such filter", }, },

    { config  = { type = "string", required = false, 
                  schema = "filters://$self.name/config_schema",
                },
    },

    --[[ other fields omitted for brevity ]]--
  },
}
```

I'm slightly unsure about pursuing this because the multiple levels of indirection mean that we _are_ asking a lot of the code that needs to resolve this reference (in order for it to be generic enough to be a native feature of the schema library). We would essentially be solving for the most complicated use-case right out of the gate.

---

**Comment by flrgh** [comment]
After a chat with @hishamhm today, a good chunk of this is going to get re-shaped, so I dismissed the pending reviews for now.

File: kong/db/schema/entities/filter_chains.lua
Comment (flrgh) [inline kong/db/schema/entities/filter_chains.lua:44]: I'm thinking it might be better to reverse this, so `config` stays the same, and we add a new field called `json_config`.

1. It would simplify this PR, because we'd no longer require a migration to rename existing `config` fields to `raw_config`.
2. Users still reap the same convenience benefit of not needing to serialize their configuration, and `config`/`json_config` is arguably a bit more intuitive.
3. It's a little bit more forward-compatible on the off chance we want to support a different serialization format. Having a choice of `config | json_config | yaml_config` is more intuitive than `config | raw_config | yaml_config`

I don't think `config`/`raw_config` is _bad_ by any stretch, but if we want to go that route we can always do the `config` => `raw_config` / `json_config` => `config` swap in a smaller PR some time in the coming weeks.

@locao @hishamhm thoughts?

Comment (flrgh) [reply kong/db/schema/entities/filter_chains.lua:44]: BTW I have a commit for this change that is passing tests locally and ready to be pushed, so it won't add any extra turnaround time if we choose to do this. :)

File: scripts/explain_manifest/fixtures/ubuntu-22.04-amd64.txt
Comment (flrgh) [inline scripts/explain_manifest/fixtures/ubuntu-22.04-amd64.txt:73]: Explanation: this feature depends on `lua-resty-ljsonschema`, which explicitly takes `lua-cjson` as a dependency, so we're now getting an additional `cjson.so` in the manifest. On my machine this file is a measly 42K, so we're not blowing up our package/image sizes.

If we want, I can put the `lua-resty-ljsonschema` addition in its own commit before merging so that it's more clear to future us why this is here.

File: kong/runloop/wasm.lua
Comment (flrgh) [inline kong/runloop/wasm.lua:584]: I think it's just about time to refactor some code out of `kong.runloop.wasm` and into a new `kong.wasm` module that is _not_ focused on runloop/proxy things, but that will be done later.

File: kong/db/schema/metaschema.lua
Comment (flrgh) [inline kong/db/schema/metaschema.lua:625]: FYI for the sake of stylistic consistency, this was copy/pasted and modified from the `subschema_key` metaschema validation code, since it serves a similar function.

File: kong/db/schema/entities/filter_chains.lua
Comment (flrgh) [reply kong/db/schema/entities/filter_chains.lua:44]: This feels like the right move for this PR; I'm making a judgment call and pushing that commit. As mentioned, I'm fine with re-introducing the change before the release if we'd like to.

**Comment by flrgh** [comment]
Rebased and cleaned up git history. From my side this is ready to go once the tests are green.

File: scripts/explain_manifest/fixtures/ubuntu-22.04-amd64.txt
Comment (flrgh) [reply scripts/explain_manifest/fixtures/ubuntu-22.04-amd64.txt:73]: Update: I _did_ put this in its own commit, but feel free to squash if that makes more sense.

File: kong/db/schema/entities/filter_chains.lua
Comment (omegabytes) [inline kong/db/schema/entities/filter_chains.lua:47]: Is there a reason we don't need `at_least_one_of` for config, json_config?

File: kong/db/schema/entities/filter_chains.lua
Comment (hishamhm) [reply kong/db/schema/entities/filter_chains.lua]: I think it's better for it to remain `config`, not `json_config`. At the level of our entity configuration, the meaning of `config` is that it is our nested configuration data structure, which happens to be validated via JSON Schema and which we happen to serialize as JSON when we pass to the filter. If we ever want to support different serialization formats, we can still use the same `config`, still validate it using JSON Schema, and then use a new field to specify the serialization format, instead of adding a new data-storage field.

As a side benefit, `config` helps pushing the JSON-based format as the "preferred default".

File: kong/db/schema/entities/filter_chains.lua
Comment (hishamhm) [reply kong/db/schema/entities/filter_chains.lua:47]: There can be filters that take no config.

Comment (flrgh) [reply kong/db/schema/entities/filter_chains.lua:47]: Yeah, configuration needs to remain optional, as it's perfectly valid to have a proxy-wasm filter that requires no configuration to function.

Specifying a schema for your filter's `json_config` using this new `my-filter.meta.json` method is one means of making it effectively required though.

File: kong/db/schema/entities/filter_chains.lua
Comment (flrgh) [reply kong/db/schema/entities/filter_chains.lua]: Sounds reasonable enough to me; I don't have a strong opinion either way. If you feel like doing some more old school gateway development you can push this through while I'm out next week. Otherwise I don't think it should be too difficult for me to get it over the inline in time for feature freeze when I am back at my desk.

**Review by locao — APPROVED** [review]

**Pull request merged**
